packages/shared/src/telemetry/collector.ts

```diff
@@ -0,0 +1,300 @@
+/**
+ * The `TelemetryCollector` class handles collection of telemetry events from Clerk SDKs. Telemetry is opt-out and can be disabled by setting a CLERK_TELEMETRY_DISABLE environment variable.
+ * The `ClerkProvider` also accepts a `telemetry` prop that will be passed to the collector during initialization:
+ *
+ * ```jsx
+ * <ClerkProvider telemetry={false}>
+ *    ...
+ * </ClerkProvider>
+ * ```
+ *
+ * For more information, please see the telemetry documentation page: https://clerk.com/docs/telemetry
+ */
+import type { InstanceType } from '@clerk/types';
+
+import { parsePublishableKey } from '../keys';
+import { isTruthy } from '../underscore';
+
+export type TelemetryCollectorOptions = {
+  /**
+   * If true, telemetry will not be collected.
+   */
+  disabled?: boolean;
+  /**
+   * If true, telemetry will not be sent, but collected events will be logged to the console.
+   */
+  debug?: boolean;
+  /**
+   * Sampling rate, 0-1
+   */
+  samplingRate?: number;
+  /**
+   * Set a custom buffer size to control how often events are sent
+   */
+  maxBufferSize?: number;
+  /**
+   * The publishableKey to associate with the collected events.
+   */
+  publishableKey?: string;
+  /**
+   * The secretKey to associate with the collected events.
+   */
+  secretKey?: string;
+  /**
+   * The current clerk-js version.
+   */
+  clerkVersion?: string;
+  /**
+   * The SDK being used, e.g. `@clerk/nextjs` or `@clerk/remix`.
+   */
+  sdk?: string;
+  /**
+   * The version of the SDK being used.
+   */
+  sdkVersion?: string;
+};
+
+type TelemetryCollectorConfig = Pick<
+  TelemetryCollectorOptions,
+  'samplingRate' | 'disabled' | 'debug' | 'maxBufferSize'
+> & {
+  endpoint: string;
+};
+
+type TelemetryMetadata = Required<
+  Pick<TelemetryCollectorOptions, 'clerkVersion' | 'sdk' | 'sdkVersion' | 'publishableKey' | 'secretKey'>
+> & {
+  /**
+   * The instance type, derived from the provided publishableKey.
+   */
+  instanceType: InstanceType;
+};
+
+type TelemetryEvent = {
+  event: string;
+  /**
+   * publishableKey
+   */
+  pk?: string;
+  /**
+   * secretKey
+   */
+  sk?: string;
+  /**
+   * instanceType
+   */
+  it: InstanceType;
+  /**
+   * clerkVersion
+   */
+  cv: string;
+  /**
+   * SDK
+   */
+  sdk?: string;
+  /**
+   * SDK Version
+   */
+  sdkv?: string;
+  payload: Record<string, string | number | boolean>;
+};
+
+const DEFAULT_CONFIG: Partial<Required<TelemetryCollectorConfig>> = {
```

Comment:
Is there any specific reason for using Partial and Required here?

Reply:
Probably not 🙃 good catch!